### PR TITLE
base stock policy

### PR DESCRIPTION
**Pull request by VictorYXL**
* Add base stock policy
    * Support load from file and from history

### CODE REVIEW

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        obj = cp.Maximize(profit)
+        prob = cp.Problem(obj, constrs)
+        prob.solve(solver = cp.GLPK_MI, verbose = True)
+        return target_stock.value
+
+    def _get_action_quantity(self, state: dict) -> int:
+        entity_id = state["entity_id"]
+        if entity_id not in self.step:
+            self.step[entity_id] = 0
+        state["step"] = self.step[entity_id]
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:60]: what is `step`? Can we update it in or_agent_state?

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        profit = cp.Variable(1)
+        buy_in.value = np.ones(time_hrz_len, dtype = np.int)
+        
+        #add constraints
+        constrs=[]
+        constrs.extend([
+            stocks >= 0, transits >= 0, sales >= 0, buy_in>=0, buy_arv>=0, buy>=0, buy[:time_hrz_len]==0,
+            stocks[current_index] == state["product_level"],
+            transits[current_index] == state["in_transition_quantity"],
+            stocks[1:time_hrz_len+1] == stocks[0:time_hrz_len] + buy_arv - sales,
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:40]: Is it necessary to formulate ticks between [0, current_index)?

File: examples/supply_chain/rl/algorithms/base_stock_policy.py (hunk 1 above)
Comment (VictorYXL) [reply examples/supply_chain/rl/algorithms/base_stock_policy.py:61]: tick=self._env.tick,

File: examples/supply_chain/rl/config.py [hunk 3]
```diff
@@ -59,6 +62,14 @@ class VehicleSelection(Enum):
     "durations": TRAIN_STEPS + EVAL_STEPS,  # number of ticks per episode
 }
 
+base_policy_conf = {
+    "data_loader": "DataLoaderFromHistory",
+    "oracle_file": "oracle_samples.csv", # Only need in DataLoaderFromFile loader
```
Comment (VictorYXL) [inline examples/supply_chain/rl/config.py:67]: remove oracle_file

File: examples/supply_chain/rl/config.py [hunk 4]
```diff
@@ -59,6 +62,14 @@ class VehicleSelection(Enum):
     "durations": TRAIN_STEPS + EVAL_STEPS,  # number of ticks per episode
 }
 
+base_policy_conf = {
+    "data_loader": "DataLoaderFromHistory",
```
Comment (VictorYXL) [inline examples/supply_chain/rl/config.py:66]: DataLoaderFromFile

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+from typing import List
+
+import cvxpy as cp
+import numpy as np
+import pandas as pd
+
+from .base_policy_data_loader import DataLoaderFromFile, DataLoaderFromHistory
+from .rule_based import RuleBasedPolicy
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:8]: from maro.rl.policy import RuleBasedPolicy

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        if self.share_same_stock_level:
+            target_stock = cp.Variable(1, integer=True)
+        else:
+            target_stock = cp.Variable(time_hrz_len, integer=True)
+
+        profit = cp.Variable(1)
+        buy_in.value = np.ones(time_hrz_len, dtype=np.int)
+
+        # add constraints
+        constrs = []
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:55]:
```
constraints = [
    stocks >= 0, ...
    ...
]
```

no need to use extend() here

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    def _get_action_quantity(self, state: dict) -> int:
+        entity_id = state["entity_id"]
+        current_tick = state["tick"]
+        self.product_level_snapshot[current_tick] = state["product_level"]
+        self.in_transit_snapshot[current_tick] = state["in_transition_quantity"]
+
+        if current_tick % self.update_frequency == 0:
+            history_start = max(current_tick - self.history_len, 0)
+            target_df = self.data_loader.load(state)
+            self.stock_quantity[entity_id] = self.calculate_stock_quantity(
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:89]: self.stock_quantity = self.calculate_stock_quantity(...)

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+                target_df,
+                self.product_level_snapshot[history_start],
+                self.in_transit_snapshot[history_start],
+                state["cur_vlt"]
+            )
+
+        if self.share_same_stock_level:
+            stock_quantity = self.stock_quantity[entity_id][0]
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:97]: ... = self.stock_quantity[0]

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        else:
+            if current_tick < self.history_len:
+                current_index = current_tick
+            else:
+                current_index = self.history_len + state["tick"] % self.update_frequency
+            stock_quantity = self.stock_quantity[entity_id][current_index]
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:103]: ... = self.stock_quantity[current_index]

File: examples/supply_chain/rl/algorithms/base_stock_policy.py (hunk 7 above)
Comment (Jinyu-W) [reply examples/supply_chain/rl/algorithms/base_stock_policy.py:89]: also let it be `self.stock_quantity: Dict[int, int] = {} `in __init__()

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
+from typing import DefaultDict, Dict, List
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:1]: remove `DefaultDict`

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+from collections import defaultdict
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:2]: `from collections ...` should be above `from typing ...`

**Review by Jinyu-W — APPROVED** [review]

**Pull request merged**